concordium-contracts-common/src/schema.rs

```diff
@@ -704,83 +711,62 @@ impl Deserial for FunctionV1 {
-                error,
-            } => {
-                out.write_u8(5)?;
-                return_value.serial(out)?;
-                error.serial(out)
-            }
-            FunctionV2::ParamRvError {
-                parameter,
-                return_value,
-                error,
-            } => {
-                out.write_u8(6)?;
-                parameter.serial(out)?;
-                return_value.serial(out)?;
-                error.serial(out)
-            }
+        let parameter = self.parameter.is_some();
+        let return_value = self.return_value.is_some();
+        let error = self.error.is_some();
+        let tag: u8 = match (parameter, return_value, error) {
+            // parameter
+            (true, false, false) => 0,
+            // return_value
+            (false, true, false) => 1,
+            // parameter + return_value
+            (true, true, false) => 2,
+            // error
+            (false, false, true) => 3,
+            // parameter + error
+            (true, false, true) => 4,
+            // return_value + error
+            (false, true, true) => 5,
+            // parameter + return_value + error
+            (true, true, true) => 6,
+            // no schema: cannot happen
+            (false, false, false) => 7,
+        };
+        out.write_u8(tag)?;
+        if let Some(p) = self.parameter.as_ref() {
+            p.serial(out)?;
+        }
+        if let Some(rv) = self.return_value.as_ref() {
+            rv.serial(out)?;
+        }
+        if let Some(err) = self.error.as_ref() {
+            err.serial(out)?;
         }
+        Ok(())
     }
 }
 
 impl Deserial for FunctionV2 {
     fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
         let idx = source.read_u8()?;
-        match idx {
-            0 => Ok(FunctionV2::Param(source.get()?)),
-            1 => Ok(FunctionV2::Rv(source.get()?)),
-            2 => Ok(FunctionV2::ParamRv {
-                parameter:    source.get()?,
-                return_value: source.get()?,
-            }),
-            3 => Ok(FunctionV2::Error(source.get()?)),
-            4 => Ok(FunctionV2::ParamError {
-                parameter: source.get()?,
-                error:     source.get()?,
-            }),
-            5 => Ok(FunctionV2::RvError {
-                return_value: source.get()?,
-                error:        source.get()?,
-            }),
-            6 => Ok(FunctionV2::ParamRvError {
-                parameter:    source.get()?,
-                return_value: source.get()?,
-                error:        source.get()?,
-            }),
-            _ => Err(ParseError::default()),
+        let mut r = FunctionV2 {
+            parameter:    None,
+            return_value: None,
+            error:        None,
+        };
+        if idx > 7 {
+            return Err(ParseError::default());
         }
+        if matches!(idx, 0 | 2 | 4 | 6) {
+            let _ = r.parameter.insert(source.get()?);
```

Comment:
Very minor thing, but I don't think you need the `let _ =` part
```suggestion
            r.parameter.insert(source.get()?);
```

Reply:
I get a warning from `clippy` about an unused return value if I remove it. That is why I kept the initial version now.

Reply:
Yes, clippy in more recent versions complains about unused option values. So this has to be there :)